### PR TITLE
[keymap] add halvves let's split keymap

### PR DESCRIPTION
**Pull request by halvves**
Hey! Just finally taking the time to toss my keymap up in the repo. Thanks!

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/lets_split/keymaps/halvves/keymap.c [hunk 1]
```diff
@@ -0,0 +1,265 @@
+#include QMK_KEYBOARD_H
+
+extern keymap_config_t keymap_config;
+
+// smooth mousekeys just put this in your config (copied from the ergo ez config)
+#define MOUSEKEY_INTERVAL 16
+#define MOUSEKEY_DELAY 0
+#define MOUSEKEY_TIME_TO_MAX 60
+#define MOUSEKEY_MAX_SPEED 7
+#define MOUSEKEY_WHEEL_DELAY 0
```
Comment (fauxpark) [inline keyboards/lets_split/keymaps/halvves/keymap.c]: These should go in your config.h instead.

File: keyboards/lets_split/keymaps/halvves/keymap.c [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+// Layer names don't all need to be of the same length, obviously, and you can also skip them
+// entirely and just use numbers.
+#define _QWERTY 0
+#define _COLEMAK 1
+#define _DVORAK 2
+#define _LOWER 3
+#define _RAISE 4
+#define _UTIL 5
+#define _MOUSE 6
+#define _ADJUST 16
```
Comment (fauxpark) [inline keyboards/lets_split/keymaps/halvves/keymap.c]: Non-consecutive layer numbers will generate extra empty layers in between, unnecessarily increasing flash usage. The better way to define layers is with an enum, eliminating the need to number them manually:

```c
enum layer_names {
    _QWERTY,
    _COLEMAK,
    _DVORAK,
    _LOWER,
    _RAISE,
    _UTIL,
    _MOUSE,
    _ADJUST
};
```

File: keyboards/lets_split/keymaps/halvves/keymap.c [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+  DVORAK,
+  LOWER,
+  RAISE,
+  UTIL,
+  MOUSE,
+};
+
+// Fillers to make layering more clear
+#define _______ KC_TRNS
+#define XXXXXXX KC_NO
```
Comment (fauxpark) [inline keyboards/lets_split/keymaps/halvves/keymap.c]: QMK defines these already, so they can be safely removed.

File: keyboards/lets_split/keymaps/halvves/keymap.c [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+ * | Shift|   Z  |   X  |   C  |   V  |   B  |   N  |   M  |   ,  |   .  |   /  |Enter |
+ * |------+------+------+------+------+------+------+------+------+------+------+------|
+ * | Util | Ctrl | Alt  | GUI  |Lower |    Space    |Raise | Mouse| Mouse|      |      |
+ * `-----------------------------------------------------------------------------------'
+ */
+[_QWERTY] = LAYOUT_ortho_4x12(
+   KC_TAB,  KC_Q,    KC_W,    KC_E,    KC_R,    KC_T,    KC_Y,    KC_U,    KC_I,    KC_O,    KC_P,    KC_BSPC, \
+   KC_ESC,  KC_A,    KC_S,    KC_D,    KC_F,    KC_G,    KC_H,    KC_J,    KC_K,    KC_L,    KC_SCLN, KC_QUOT, \
+   KC_LSFT, KC_Z,    KC_X,    KC_C,    KC_V,    KC_B,    KC_N,    KC_M,    KC_COMM, KC_DOT,  KC_SLSH, KC_ENT , \
+   UTIL,    KC_LCTL, KC_LALT, KC_LGUI, LOWER,   KC_SPC,  KC_SPC,  RAISE,   MOUSE,   MOUSE,   XXXXXXX, XXXXXXX  \
```
Comment (fauxpark) [inline keyboards/lets_split/keymaps/halvves/keymap.c]: The backslashes at the end of these lines are not necessary. You can keep them if you want as it isn't hurting anything, but IMO it looks neater without them.

File: keyboards/lets_split/keymaps/halvves/rules.mk [hunk 5]
```diff
@@ -0,0 +1,4 @@
+ifndef QUANTUM_DIR
+	include ../../../../Makefile
+endif
```
Comment (fauxpark) [inline keyboards/lets_split/keymaps/halvves/rules.mk]: This block should be removed.

File: keyboards/lets_split/keymaps/halvves/keymap.c (hunk 3 above)
Comment (halvves) [reply keyboards/lets_split/keymaps/halvves/keymap.c]: Sorry about all of this. I made this keymap a while back based on (I think) the default keymap and haven't updated in a minute. 😔

File: keyboards/lets_split/keymaps/halvves/keymap.c [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+ * | Tab  |   Q  |   W  |   F  |   P  |   G  |   J  |   L  |   U  |   Y  |   ;  | Bksp |
+ * |------+------+------+------+------+-------------+------+------+------+------+------|
+ * | Esc  |   A  |   R  |   S  |   T  |   D  |   H  |   N  |   E  |   I  |   O  |  '   |
+ * |------+------+------+------+------+------|------+------+------+------+------+------|
+ * | Shift|   Z  |   X  |   C  |   V  |   B  |   K  |   M  |   ,  |   .  |   /  |Enter |
+ * |------+------+------+------+------+------+------+------+------+------+------+------|
+ * | Util | Ctrl | Alt  | GUI  |Lower |Space |Space |Raise | Mouse| Mouse|      |      |
+ * `-----------------------------------------------------------------------------------'
+ */
+[_COLEMAK] = LAYOUT_ortho_4x12( \
```
Comment (fauxpark) [inline keyboards/lets_split/keymaps/halvves/keymap.c]:
```suggestion
[_COLEMAK] = LAYOUT_ortho_4x12(
```

File: keyboards/lets_split/keymaps/halvves/keymap.c [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+ * | Tab  |   '  |   ,  |   .  |   P  |   Y  |   F  |   G  |   C  |   R  |   L  | Bksp |
+ * |------+------+------+------+------+-------------+------+------+------+------+------|
+ * | Esc  |   A  |   O  |   E  |   U  |   I  |   D  |   H  |   T  |   N  |   S  |  /   |
+ * |------+------+------+------+------+------|------+------+------+------+------+------|
+ * | Shift|   ;  |   Q  |   J  |   K  |   X  |   B  |   M  |   W  |   V  |   Z  |Enter |
+ * |------+------+------+------+------+------+------+------+------+------+------+------|
+ * | Util | Ctrl | Alt  | GUI  |Lower |Space |Space |Raise | Mouse| Mouse|      |      |
+ * `-----------------------------------------------------------------------------------'
+ */
+[_DVORAK] = LAYOUT_ortho_4x12( \
```
Comment (fauxpark) [inline keyboards/lets_split/keymaps/halvves/keymap.c]:
```suggestion
[_DVORAK] = LAYOUT_ortho_4x12(
```

File: keyboards/lets_split/keymaps/halvves/keymap.c [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+ * |   ~  |   !  |   @  |   #  |   $  |   %  |   ^  |   &  |   *  |   (  |   )  | Del  |
+ * |------+------+------+------+------+-------------+------+------+------+------+------|
+ * | Del  |  F1  |  F2  |  F3  |  F4  |  F5  |  F6  |   _  |   +  |   {  |   }  |  |   |
+ * |------+------+------+------+------+------|------+------+------+------+------+------|
+ * |      |  F7  |  F8  |  F9  |  F10 |  F11 |  F12 |ISO ~ |ISO | |      |      |      |
+ * |------+------+------+------+------+------+------+------+------+------+------+------|
+ * |      |      |      |      |      |             |      | Next | Vol- | Vol+ | Play |
+ * `-----------------------------------------------------------------------------------'
+ */
+[_LOWER] = LAYOUT_ortho_4x12( \
```
Comment (fauxpark) [inline keyboards/lets_split/keymaps/halvves/keymap.c]:
```suggestion
[_LOWER] = LAYOUT_ortho_4x12(
```

File: keyboards/lets_split/keymaps/halvves/keymap.c [hunk 9]
```diff
@@ -0,0 +1,252 @@
...
+ * |   `  |   1  |   2  |   3  |   4  |   5  |   6  |   7  |   8  |   9  |   0  | Del  |
+ * |------+------+------+------+------+-------------+------+------+------+------+------|
+ * | Del  |  F1  |  F2  |  F3  |  F4  |  F5  |  F6  |   -  |   =  |   [  |   ]  |  \   |
+ * |------+------+------+------+------+------|------+------+------+------+------+------|
+ * |      |  F7  |  F8  |  F9  |  F10 |  F11 |  F12 |ISO # |ISO / |      |      |      |
+ * |------+------+------+------+------+------+------+------+------+------+------+------|
+ * |      |      |      |      |      |             |      | Next | Vol- | Vol+ | Play |
+ * `-----------------------------------------------------------------------------------'
+ */
+[_RAISE] = LAYOUT_ortho_4x12( \
```
Comment (fauxpark) [inline keyboards/lets_split/keymaps/halvves/keymap.c]:
```suggestion
[_RAISE] = LAYOUT_ortho_4x12(
```

File: keyboards/lets_split/keymaps/halvves/keymap.c [hunk 10]
```diff
@@ -0,0 +1,252 @@
...
+ * |      |      |      |      |      |      |      |      |      |      |      |      |
+ * |------+------+------+------+------+-------------+------+------+------+------+------|
+ * |      |      |      |      |      |      | LEFT | DOWN |  UP  | RGHT |      |      |
+ * |------+------+------+------+------+------|------+------+------+------+------+------|
+ * |      |      |      |      |      |      |      |      |      |      |      |      |
+ * |------+------+------+------+------+------+------+------+------+------+------+------|
+ * |      |      |      |      |      |      |      |      |      |      |      |      |
+ * `-----------------------------------------------------------------------------------'
+ */
+[_UTIL] = LAYOUT_ortho_4x12( \
```
Comment (fauxpark) [inline keyboards/lets_split/keymaps/halvves/keymap.c]:
```suggestion
[_UTIL] = LAYOUT_ortho_4x12(
```

File: keyboards/lets_split/keymaps/halvves/keymap.c [hunk 11]
```diff
@@ -0,0 +1,252 @@
...
+ * |      |      | M Up |      |      |      |      |      |      |L Clk |R Clk | W Up |
+ * |------+------+------+------+------+-------------+------+------+------+------+------|
+ * |      |M Left|M Down|M Rght|      |      |      |      |      |      |      |W Down|
+ * |------+------+------+------+------+------|------+------+------+------+------+------|
+ * |      |      |      |      |      |      |      |      |      |      |      |      |
+ * |------+------+------+------+------+------+------+------+------+------+------+------|
+ * |      |      |      |      |      |      |      |      |      |      |      |      |
+ * `-----------------------------------------------------------------------------------'
+ */
+[_MOUSE] = LAYOUT_ortho_4x12( \
```
Comment (fauxpark) [inline keyboards/lets_split/keymaps/halvves/keymap.c]:
```suggestion
[_MOUSE] = LAYOUT_ortho_4x12(
```

File: keyboards/lets_split/keymaps/halvves/keymap.c [hunk 12]
```diff
@@ -0,0 +1,252 @@
...
+ * |      | Reset|      |      |      |      |      |      |      |      |      |  Del |
+ * |------+------+------+------+------+-------------+------+------+------+------+------|
+ * |      |      |      |Aud on|Audoff|AGnorm|AGswap|Qwerty|Colemk|Dvorak|      |      |
+ * |------+------+------+------+------+------|------+------+------+------+------+------|
+ * |      |      |      |      |      |      |      |      |      |      |      |      |
+ * |------+------+------+------+------+------+------+------+------+------+------+------|
+ * |      |      |      |      |      |             |      |      |      |      |      |
+ * `-----------------------------------------------------------------------------------'
+ */
+[_ADJUST] =  LAYOUT_ortho_4x12( \
```
Comment (fauxpark) [inline keyboards/lets_split/keymaps/halvves/keymap.c]:
```suggestion
[_ADJUST] =  LAYOUT_ortho_4x12(
```

File: keyboards/lets_split/keymaps/halvves/keymap.c [hunk 13]
```diff
@@ -0,0 +1,252 @@
...
+}
+
+bool process_record_user(uint16_t keycode, keyrecord_t *record) {
+  switch (keycode) {
+    case QWERTY:
+      if (record->event.pressed) {
+        #ifdef AUDIO_ENABLE
+          PLAY_SONG(tone_qwerty);
+        #endif
+        persistent_default_layer_set(1UL<<_QWERTY);
```
Comment (fauxpark) [inline keyboards/lets_split/keymaps/halvves/keymap.c]: These I think should be `set_single_persistent_default_layer(_QWERTY)` etc, without the `1UL<<`.

File: keyboards/lets_split/keymaps/halvves/keymap.c [hunk 14]
```diff
@@ -0,0 +1,252 @@
+#include QMK_KEYBOARD_H
+
+extern keymap_config_t keymap_config;
```
Comment (fauxpark) [inline keyboards/lets_split/keymaps/halvves/keymap.c]:
```suggestion
```

File: keyboards/lets_split/keymaps/halvves/keymap.c [hunk 15]
```diff
@@ -0,0 +1,252 @@
...
+#ifdef AUDIO_ENABLE
+float tone_qwerty[][2]     = SONG(QWERTY_SOUND);
+float tone_dvorak[][2]     = SONG(DVORAK_SOUND);
+float tone_colemak[][2]    = SONG(COLEMAK_SOUND);
+#endif
+
+void persistent_default_layer_set(uint16_t default_layer) {
+  eeconfig_update_default_layer(default_layer);
+  default_layer_set(default_layer);
+}
```
Comment (drashna) [inline keyboards/lets_split/keymaps/halvves/keymap.c]: Could you remove this function?  

It's been deprecated in favor of the `set_single_persistent_default_layer` function (which is a core function).

https://docs.qmk.fm/#/ref_functions?id=setting-the-persistent-default-layer

Comment (halvves) [reply keyboards/lets_split/keymaps/halvves/keymap.c]: Oh nice that this is core now. Thanks!

File: keyboards/lets_split/keymaps/halvves/keymap.c [hunk 16]
```diff
@@ -0,0 +1,196 @@
...
+  _______, RESET,   _______, _______, _______, _______, _______, _______, _______, _______, _______, KC_DEL,
+  _______, _______, _______, AU_ON,   AU_OFF,  AG_NORM, AG_SWAP, QWERTY,  COLEMAK, DVORAK,  _______, _______,
+  _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,
+  _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______
+)
+
+
+};
+
+uint32_t layer_state_set_user(uint32_t state) {
```
Comment (drashna) [inline keyboards/lets_split/keymaps/halvves/keymap.c]: Sorry, one more thing (I promise): 
```suggestion
layer_state_t layer_state_set_user(layer_state_t state) {
```

We've moved from a hard coded 32 bit mask to a definable mask. So, if you could change this?

**Pull request merged**